Avoid comparing full objects during update to optimize performance

tsaarni opened the pull request:
This change optimizes processing of `HTTPProxy`, `Secret` and other objects by avoiding full object comparison. This reduces CPU usage during object updates.

Comparing the full objects with `cmp.Equal()` during `EventHandler.onUpdate()` was very costly, which was especially visible in big clusters with a lot of objects. In general, the load caused by `cmp.Equal()` during update was overwhelming compared to the load that Contour otherwise generates in total.

For objects that implement `generation` it can be used, instead of comparing the full object. Generation is updated only when object spec is updated. Updates to `status` or `annotations` fields will not trigger new generation, which needs to be considered separately for objects where those changes matter.

Fixes #5024.

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

tsaarni commented:
Notes for reviewers:

- Tests that update objects will now need unique `generation` to trigger update. Those test cases have been updated by calling a function that gives unique `generation`.
- There remains almost 1500 places in test code that still constructs `ObjectMeta` in-place and will have `generation: 0`. I left them untouched since replacing them requires a lot of manual work.
- Status updates still use `cmp.Equal()` ([link](https://github.com/projectcontour/contour/blob/main/internal/k8s/helpers.go)) and there doesn't seem to be straightforward way to avoid comparing the complete object. However, according to profiling this does not seem to be a big issue even for clusters with many objects.
- As mentioned in #5024, the Overview text in [go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp) discourages its use for production.

skriss commented on internal/contour/handler.go:
```diff
@@ -189,19 +183,12 @@ func (e *EventHandler) onUpdate(op interface{}) bool {
-			cmpopts.IgnoreFields(metav1.ObjectMeta{}, "ManagedFields"),
-		) {
+		// Ignore if generation is the same.
+		if op.oldObj.(client.Object).GetGeneration() == op.newObj.(client.Object).GetGeneration() {
```
I'm observing that in the old code, we did not ignore status for things like Services, Secrets, etc. Wondering if there's any particular need to trigger DAG rebuilds when status for those referenced resources changes? The only thing I can think of off the top of my head is the Envoy service's status.loadBalancer changing which should trigger an update to Ingress/HTTPProxy status, but that goes through a different code path than this.

tsaarni replied:
True, it was unfortunately not simple as this. Not all objects have generation field set at all.

skriss reviewed (commented):
Thanks @tsaarni!

tsaarni commented:
After some digging, I've ended up enumerating each type and associated comparison method. For example, all CRDs can be compared by `generation` field, but for some others I used `resourceVersion`. I'm not sure if this makes sense either. The unit tests are not that useful here since the objects are crafted by hand. Preferably updates would be all tested in e2e environment.

Since I now enumerate types, I added default/fallback implementation - this is to catch corner cases where new types are added later but update for `IsObjectEqual()` was missed. The intention is that a full comparison is made in that case.

I experimented with `reflect.DeepEquals()` and compared the performance to `cmp.Equal()`. Curiously, it seemed to me that `reflect.DeepEquals()` does not suffer from the performance problem, even if I let it process thousands of `Secrets` by full comparison. This requires more investigation. Maybe in reality it did not make full comparison in my test...

tsaarni commented:
For the curious, here is benchmarking:


```go

const numBytes = 5000

func BenchmarkCmpEqual(b *testing.B) {
	a := v1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name: "test",
		},
		Data: map[string][]byte{
			"test": make([]byte, numBytes),
		},
	}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		cmp.Equal(&a, &a)
	}
}

func BenchmarkDeepEqual(b *testing.B) {
	a := v1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name: "test",
		},
		Data: map[string][]byte{
			"test": make([]byte, numBytes),
		},
	}
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		reflect.DeepEqual(&a, &a)
	}
}
```

Results:

```
# numBytes = 5000
BenchmarkCmpEqual-20     	     430	   2964049 ns/op	  319721 B/op	     168 allocs/op
BenchmarkDeepEqual-20    	27470511	        40.72 ns/op	       0 B/op	       0 allocs/op

# numBytes = 50000
BenchmarkCmpEqual-20     	      40	  28284435 ns/op	 4606866 B/op	     210 allocs/op
BenchmarkDeepEqual-20    	28174531	        40.81 ns/op	       0 B/op	       0 allocs/op
```

For 5000-byte secret, `cmp.Equal()` took 2.96ms, for 50000-byte secret it took 10 times that. An example of a big secret (50000 bytes and waaay above) is for example helm manifest secrets of type `type: helm.sh/release.v1`.  When installing via helm, it will update the secret which will trigger considerable CPU peak right at the startup.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +87,60 @@ func isStatusEqual(objA, objB interface{}) bool {
+	case *gatewayapi_v1alpha2.TLSRoute:
+		return isGenerationEqual(a, b)
+
+	// Fallback implementation: compare the full object if not handled above.
```
I'm thinking we could probably just use ResourceVersion comparison as the fallback strategy for any resources that Contour is not updating the status of, WDYT? Would be nice and simple. I don't really see a need to deep-compare.

tsaarni replied:
Yes, I agree.

tsaarni replied:
> I'm thinking we could probably just use ResourceVersion comparison as the fallback strategy for any resources that Contour is not updating the status of, WDYT? Would be nice and simple. I don't really see a need to deep-compare.

Since it is just fallback, it would be OK even though [`ManagedFields`](https://kubernetes.io/docs/reference/using-api/server-side-apply/#field-management) had caused performance issues previously https://github.com/projectcontour/contour/pull/2987?

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +87,60 @@ func isStatusEqual(objA, objB interface{}) bool {
+	case *v1.Endpoints:
+		return isResourceVersionEqual(a, b)
+
+	case *networking_v1.Ingress:
+		return isGenerationEqual(a, b)
```
Ingress seems like the only non-CRD here; does it implement Generation?

tsaarni commented:
I'm still missing filtering of [`ManagedFields`](https://kubernetes.io/docs/reference/using-api/server-side-apply/#field-management) which had caused performance issues previously https://github.com/projectcontour/contour/pull/2987.

I was also thinking that for the types that have only `ResourceVersion`, we could use that for "fast path" decision, taking care of the periodic controller-runtime sync, but then make a full comparison of e.g. `Secret.Data` with `reflect.DeepEqual()` to filter out e.g. annotation/label changes etc?  

```go
return isResourceVersionEqual(a, b) || reflect.DeepEqual(a.(*v1.Secret).Data, b.(*v1.Secret).Data)
```

tsaarni commented:
I've added some fine-grained comparin logic that should approximate the previous behavior, with the exception that "fast path" is taken when possible.  For "slow path" I changed to `reflect.DeepEquals()` which is not that bad.

I've added unit tests (not e2e test) but some are based on manifests dumped from real cluster. I'm not sure if that is worth it or not, since we can of course craft those objects by hand as well. 

I have not replaced `cmp.Equal()` from [`k8s.isStatusEqual()`](https://github.com/projectcontour/contour/blob/d044dbd7b72587f4e40052af13f76439f25ed363/internal/k8s/helpers.go#L29) but it may be preferable to remove that too.

It would be also nice to investigate filtering `Secrets` of other types than `Opaque` and `kubernetes.io/tls` no controller-runtime level, so that e.g. huge helm manifest secrets `helm.sh/release.v1` would be dropped as early as possible.

sunjayBhatia reviewed (commented):
If possible could we pull the `ObjectMeta` and other helper changes into their own PR? just in case we want to backport or revert etc. the specific changes around improving the object comparison bits

tsaarni commented:
> If possible could we pull the `ObjectMeta` and other helper changes into their own PR? just in case we want to backport or revert etc. the specific changes around improving the object comparison bits

I created https://github.com/projectcontour/contour/pull/5094 and rebased this PR on top of that. I hope github does not get confused :)

tsaarni commented:
I've now looked at the remaining use of `cmp.Equal()` which we have in [`k8s.isStatusEqual()`](https://github.com/projectcontour/contour/blob/d044dbd7b72587f4e40052af13f76439f25ed363/internal/k8s/helpers.go#L29).  I'm proposing leaving that untouched for now, even if use of `cmp.Equal()` is not recommended in production environment according to the [package description](https://pkg.go.dev/github.com/google/go-cmp/cmp).

Reasoning:
* It does not seem to cause any considerable CPU churn, not in any scenario that I managed to simulate. 
* Ignoring `LastTransitionTime` fields in various objects looks complicated to me unless it could be done like [here](https://github.com/kubernetes/kubernetes/blob/b0ed87078e4042c12e456c8427dee0fb947a724c/pkg/kubelet/kubelet_node_status.go#L811-L816) in similar code in Kubernetes, simply by ignoring `Conditions` completely?

If you have ideas around this, I'd gladly work on it. Otherwise I'm concluding this PR.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
+
+	switch old := old.(type) {
+
+	// Fast path for objects that implement Generation and where only spec changes matter.
```
This seems like a reasonable comparison for this set of resources given that status/annotation/label changes shouldn't need to trigger rebuilds for them.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
+	case *contour_api_v1alpha1.ContourConfiguration:
+		return isGenerationEqual(old, new), nil
+
+	case *gatewayapi_v1beta1.GatewayClass:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.Gateway:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.HTTPRoute:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1alpha2.TLSRoute:
+		return isGenerationEqual(old, new), nil
```
Need to add `*gatewayapi_v1beta1.ReferenceGrant` here as well

tsaarni replied:
Fixed!

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
+	// Slow path: compare the content of the objects.
+	case *contour_api_v1.HTTPProxy:
+		if new, ok := new.(*contour_api_v1.HTTPProxy); ok {
+			return isGenerationEqual(old, new) &&
+				apiequality.Semantic.DeepEqual(old.ObjectMeta.Annotations, new.ObjectMeta.Annotations), nil
```
I _think_ this is reasonable comparison logic, though worth noting that previously, Label changes (for example) would trigger a DAG rebuild whereas now they won't. I can't think of a reason that they would need to trigger a rebuild, but it is a slight change in behavior. Similarly for some of the below objects.

tsaarni replied:
I can add more fields, but it seems that only annotations would get used by Contour. And e.g. object name is immutable, so it is safe to exclude.

Adding everything is bit problematic since `DeepEqual()` and alikes work in opposite way to `cmp.Equal()` which gave the option to ignore named fields.

skriss replied:
I think it's OK as is; just need to squint at each individual object to make sure we're comparing the full set of things.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
+
+	}
+
+	// ResourseVersions are not equal and we don't know how to compare the object type.
+	return false, fmt.Errorf("do not know how to compare %s", old.(client.Object).GetObjectKind().GroupVersionKind().String())
```
Do you think it's necessary to return an error here? We're effectively falling back to ResourceVersion comparison, which seems fine. I guess the question is, does logging this have any benefit for the user/operator?

tsaarni replied:
I was thinking how to alert the developer who might miss adding detailed comparison when introducing new type. I wonder, can it be said that for every handled type, there should be explicit comparison? 

If yes, then without error printout, it might be likely that we only notice the problem much later e.g. when performance impact becomes inevitable.

If no, then I can remove the error.

skriss replied:
> I was thinking how to alert the developer who might miss adding detailed comparison when introducing new type. I wonder, can it be said that for every handled type, there should be explicit comparison?

I think this is probably right. In that case, I wonder if it'd be better to use a panic on an unknown object type? This would cause E2E tests to fail, which would then force the code to be updated before getting merged, whereas the error log could always be missed.

As an illustrative example, I know we need to add Namespace and GRPCRoute to this function still, but they are only used for Gateway API so you wouldn't even see the errors in the log unless you were testing with Gateway API. But, with a panic in place, the E2E's would fail.

I know intentionally adding `panic()` to code doesn't sit right with everyone, so curious what folks think.

Another option would be to look to enforce this at compile time, by introducing an interface that all types we cache would be required to implement. I'd like to look into this some more to see if/how it could work, but it doesn't need to be done as part of this PR.

sunjayBhatia replied:
> Another option would be to look to enforce this at compile time, by introducing an interface that all types we cache would be required to implement. I'd like to look into this some more to see if/how it could work, but it doesn't need to be done as part of this PR.

yeah this is sorta what i was heading towards with the generics approach

we would probably need to add some wrapper types of our own around the k8s core resources or GW API resources if we need to get at some of the internals (spec field, status field, etc.) with a `.Equal` method or something similar and make that interface a constraint

the way i did it doesn't go this far ^ and still has switches on types in the comparison helper, so could be done better with some more thinking

skriss replied:
@sunjayBhatia yeah I like that idea. I'm fine with the impl as-is for now as long as we're looking into some improvements here down the road, since I do think it's super-easy to miss all the various spots to update for new types right now. But, in this case, falling back to ResourceVersion comparison isn't so bad either.

tsaarni replied:
Thank You @sunjayBhatia and @skriss for looking into this!

I must confess that I have not looked much into Go generics yet, so I'm likely not knowledgeable enough to grasp how it could be done :) To me it seems impossible to influence the interfaces in such a way that the problem is converted into compiler time error, because the objects are Kubernetes types, constructed by controller-runtime.

sunjayBhatia replied:
yeah for sure, the controller-runtime/client-go controller/reconciler/handler interfaces all use `interface{}` there will likely be at the very least 1 place we need to do a type assertion

skriss reviewed (commented):
Thanks @tsaarni, I'm in favor of this change in general, just being paranoid about any behavior changes given the importance of this logic.

sunjayBhatia reviewed (commented):
this looks great!

I think in some places we could maybe do away with the inline type assertions etc. and also ensure consistent types between `old` and `new` objects with type constraints etc.

I did a little spike on using generics on top of your changes here: https://github.com/sunjayBhatia/contour/commit/92c3b9ecec4f72d1bcc5af988a1da351b5646bdd (still has a hack to use the type assertion switch to get at the internal fields)

let me know what you think (this doesn't necessarily have to be incorporated in this PR)

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,81 @@ func isStatusEqual(objA, objB interface{}) bool {
+		return true, nil
+	}
+
+	switch old := old.(type) {
```
need to add handling for `Namespace` and `GRPCRoute`. In the case of Namespace, the Gateway API processor actually looks at the labels, so those need to taken into account for the comparison. GRPCRoute _should_ be like the other two Gateway API route types.

tsaarni replied:
Thank you for spotting this, I've now added these types!

tsaarni commented:
>I think in some places we could maybe do away with the inline type assertions etc. and also ensure consistent types between old and new objects with type constraints etc.
>
> I did a little spike on using generics on top of your changes here: [sunjayBhatia@92c3b9e](https://github.com/sunjayBhatia/contour/commit/92c3b9ecec4f72d1bcc5af988a1da351b5646bdd) (still has a hack to use the type assertion switch to get at the internal fields)

Thank you @sunjayBhatia! 

You added great cleanups that I've now included! But I did not understand what the generic `IsObjectEqual()` is adding, if we still need to have the switch-case to compare relevant bits of each type? It does not look very generic :sweat_smile:

BTW I assume we could just trust `old` and `new` having consistent type, even without checking?

tsaarni commented:
> I wonder if it'd be better to use a panic on an unknown object type?
>...
> falling back to ResourceVersion comparison isn't so bad either.

@skriss I'm torn between these two :thinking:  

Panic would be the correct approach in my mind. No one will read error printouts from e2e tests, and they are not automatically causing test cases to fail either.  As you pointed out, I already missed some types.  But do I then dare to add it with the risk of panicking in production? :-)

skriss commented:
> > I wonder if it'd be better to use a panic on an unknown object type?
> > ...
> > falling back to ResourceVersion comparison isn't so bad either.
> 
> @skriss I'm torn between these two 🤔
> 
> Panic would be the correct approach in my mind. No one will read error printouts from e2e tests, and they are not automatically causing test cases to fail either. As you pointed out, I already missed some types. But do I then dare to add it with the risk of panicking in production? :-)

Yeah, I feel similarly. I think let's hold off on the panic for now; @sunjayBhatia and I discussed some ideas about at least minimizing the number of places where we have to do type-switching, and then using some wrapper structs + interfaces in other places, which I think might provide the same benefit without the risk of somehow causing a panic in production. We can look into that as a followup. In the mean time, we'll just have to remember to update this for any new types.

tsaarni commented:
I guess one option could be also to look closer into `cmp.Equal()` and `apiequality.Semantic.DeepEqual()` and try figure out if it is possible to have comparison that combines the flexibility of the former with the speed of the latter.

PS: Sorry, I had some freak accident with the github GUI and ended up editing your previous response although I'm pretty sure I did not go through edit menu :open_mouth: I did edit it back to original, since github does not allow deleting latest edit.

sunjayBhatia commented:
> > I think in some places we could maybe do away with the inline type assertions etc. and also ensure consistent types between old and new objects with type constraints etc.
> > I did a little spike on using generics on top of your changes here: [sunjayBhatia@92c3b9e](https://github.com/sunjayBhatia/contour/commit/92c3b9ecec4f72d1bcc5af988a1da351b5646bdd) (still has a hack to use the type assertion switch to get at the internal fields)
> 
> Thank you @sunjayBhatia!
> 
> You added great cleanups that I've now included! But I did not understand what the generic `IsObjectEqual()` is adding, if we still need to have the switch-case to compare relevant bits of each type? It does not look very generic 😅
> 
> BTW I assume we could just trust `old` and `new` having consistent type, even without checking?

yeah we likely need to add wrapper types around the resource types to make things fully generic, esp. to make sure we don't have to do the type assertions you mention to get at the internals of the struct fields

i have some ideas around this area which will require wider changes so no sweat to disregard the mish-mash of generics additions

skriss commented on internal/contour/handler.go:
```diff
@@ -189,23 +183,26 @@ func (e *EventHandler) onUpdate(op interface{}) bool {
+				e.WithError(err).WithField("op", "update").Errorf("%T error comparing objects", op.newObj)
+			}
+			if equal {
+				e.WithField("op", "update").Debugf("%T skipping update, only status has changed", op.newObj)
```
It'd be nice to:
- log the namespace/name of the resource here as well, for debugging purposes
- adjust the message to be a little more general since it's not 100% accurate -- something like "skipping update, no changes to relevant fields"

tsaarni replied:
I've added this. 

I also noticed that `obj.GetObjectKind().GroupVersionKind()` is (mostly) empty due to [kubernetes/client-go#541](https://www.github.com/kubernetes/client-go/issues/541) so I switched to printing type of the object.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,73 @@ func isStatusEqual(objA, objB interface{}) bool {
+		if new, ok := new.(*v1.Namespace); ok {
+			return apiequality.Semantic.DeepEqual(old.Labels, new.Labels), nil
+		}
+
+	default:
+
```
This can be dropped.

tsaarni replied:
Fixed.

skriss commented on internal/k8s/helpers.go:
```diff
@@ -83,3 +89,73 @@ func isStatusEqual(objA, objB interface{}) bool {
+
+	}
+
+	// ResourseVersions are not equal and we don't know how to compare the object type.
```
nit:
```suggestion
	// ResourceVersions are not equal and we don't know how to compare the object type.
```

tsaarni replied:
Fixed

skriss reviewed (changes requested):
Couple last comments then I think I am comfortable with going ahead with this, thanks @tsaarni!

skriss reviewed (approved):
LGTM, thanks @tsaarni!

Will leave for @sunjayBhatia to take another look as well.

sunjayBhatia commented:
thanks for the great work @tsaarni!

ravilr commented:
Related to https://github.com/projectcontour/contour/pull/4788#discussion_r1129065936

Did anything change with respect to object comparisons b/w contour release v1.22.x and v1.24.1 ? 

we're noticing increased CPU usage (in addition to memory usage) after upgrading to v1.24.1 and cpu profile shows `google/go-cmp` methods in the top samples:

```
$ go tool pprof contourcpu.pprof 
File: contour
Type: cpu
Time: Mar 8, 2023 at 12:31am (PST)
Duration: 120.17s, Total samples = 136.15s (113.30%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top 20
Showing nodes accounting for 101.59s, 74.62% of 136.15s total
Dropped 482 nodes (cum <= 0.68s)
Showing top 20 nodes out of 66
      flat  flat%   sum%        cum   cum%
    20.06s 14.73% 14.73%     52.57s 38.61%  github.com/google/go-cmp/cmp/cmpopts.structFilter.filter
    14.08s 10.34% 25.08%     30.12s 22.12%  reflect.(*rtype).AssignableTo
    13.69s 10.06% 35.13%     75.38s 55.37%  github.com/google/go-cmp/cmp.Options.filter
     9.04s  6.64% 41.77%     61.61s 45.25%  github.com/google/go-cmp/cmp.pathFilter.filter
     6.09s  4.47% 46.24%      6.15s  4.52%  reflect.implements
     5.76s  4.23% 50.47%      9.89s  7.26%  reflect.directlyAssignable
     4.86s  3.57% 54.04%      4.86s  3.57%  aeshashbody
     4.52s  3.32% 57.36%    119.68s 87.90%  github.com/google/go-cmp/cmp.(*state).compareAny
     3.27s  2.40% 59.76%     14.09s 10.35%  runtime.gcDrain
     2.57s  1.89% 61.65%      9.53s  7.00%  runtime.mapdelete
     2.38s  1.75% 63.40%      2.38s  1.75%  reflect.(*rtype).hasName (inline)
     2.36s  1.73% 65.13%      7.06s  5.19%  runtime.mapassign
     2.16s  1.59% 66.72%      2.16s  1.59%  reflect.(*rtype).Kind (partial-inline)
     2.01s  1.48% 68.20%      8.41s  6.18%  runtime.scanobject
     1.98s  1.45% 69.65%      2.02s  1.48%  runtime.pageIndexOf (inline)
     1.71s  1.26% 70.91%      1.71s  1.26%  runtime.memhash64
     1.38s  1.01% 71.92%      4.98s  3.66%  runtime.mapaccess2
     1.28s  0.94% 72.86%      1.57s  1.15%  runtime.findObject
     1.24s  0.91% 73.77%     76.86s 56.45%  github.com/google/go-cmp/cmp.(*state).tryOptions
     1.15s  0.84% 74.62%      2.86s  2.10%  runtime.typehash
```

tsaarni commented:
> Did anything change with respect to object comparisons b/w contour release v1.22.x and v1.24.1 ?

@ravilr Hmm, I cannot spot anything specific with a quick glance, but there is a lot of changed code. At least some new resource types seem to have been introduced, mainly related to Gateway API. The comparison was based on the same go-cmp library even previously, but the library version has changed too. I did not compare different versions when benchmarking.
